Haskell extraction generates invalid Haskell

### Description of the problem

[Passing version from 3 weeks ago](https://github.com/mit-plv/fiat-crypto/actions/runs/14977402102/job/42078833780) (de8e6e97603ac2853a88438192821491a8b3163e):

[ExtractionHaskell-source-master-de8e6e97603ac2853a88438192821491a8b3163e-passing.zip](https://github.com/user-attachments/files/20596554/ExtractionHaskell-source-master-de8e6e97603ac2853a88438192821491a8b3163e-passing.zip)

[Failing version](https://github.com/mit-plv/fiat-crypto/actions/runs/15369438942/job/43248316649) (8c1f088465175e27a87dd6dd38c063cdc53a1a8c):
[ExtractionHaskell-source-master-8c1f088465175e27a87dd6dd38c063cdc53a1a8c-failing.zip](https://github.com/user-attachments/files/20596555/ExtractionHaskell-source-master-8c1f088465175e27a87dd6dd38c063cdc53a1a8c-failing.zip)

GHC version: The Glorious Glasgow Haskell Compilation System, version 9.12.2

Error:
```
command time -f "src/ExtractionHaskell/fiat_crypto (real: %e, user: %U, sys: %S, mem: %M ko)" ghc +RTS -M9G -RTS -o src/ExtractionHaskell/fiat_crypto src/ExtractionHaskell/fiat_crypto.hs
[1 of 2] Compiling Main             ( src/ExtractionHaskell/fiat_crypto.hs, src/ExtractionHaskell/fiat_crypto.o )
src/ExtractionHaskell/fiat_crypto.hs:12863:68: error: [GHC-76037]
    Not in scope:
      type constructor or class ‘With_unif_rewrite_ruleTP_gen’
      |
12863 |    Build_rewrite_rule_data Prelude.Bool Prelude.Bool Prelude.Bool (With_unif_rewrite_ruleTP_gen
      |                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/ExtractionHaskell/fiat_crypto.hs:75828:2: error: [GHC-76037]
    Not in scope: type constructor or class ‘Assembly_output_first_opt’
      |
75828 |  Assembly_output_first_opt Assembly_argument_registers_left_to_right_opt 
      |  ^^^^^^^^^^^^^^^^^^^^^^^^^

src/ExtractionHaskell/fiat_crypto.hs:75828:28: error: [GHC-76037]
    Not in scope:
      type constructor or class ‘Assembly_argument_registers_left_to_right_opt’
      |
75828 |  Assembly_output_first_opt Assembly_argument_registers_left_to_right_opt 
      |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Command exited with non-zero status 1
src/ExtractionHaskell/fiat_crypto (real: 18.16, user: 15.92, sys: 2.19, mem: 4468376 ko)
```

### Small Rocq / Coq file to reproduce the bug

This can be tested on Rocq's CI (assuming ghc is installed) with

```bash
#!/usr/bin/env bash

set -e

# fiat-crypto job sets up the sources
if [ "$DOWNLOAD_ONLY" ]; then exit 0; fi

ci_dir="$(dirname "$0")"
. "${ci_dir}/ci-common.sh"

# Regardless of where the dependencies came from when building
# ci-fiat_crypto, we don't need them for building the Haskell
# binaries, so we use all external dependencies.
make_args=(EXTERNAL_DEPENDENCIES=1)

( cd "${CI_BUILD_DIR}/fiat_crypto"
  make "${make_args[@]}" -j 1 standalone-haskell
)
```

Unfortunately this target takes ~46m.  Building target `src/ExtractionHaskell/fiat_crypto` will take only 2--4 minutes, and we can provide a `lite-standalone-haskell` target that builds only that if desired.

### Version of Rocq / Coq where this bug occurs

8c1f088465175e27a87dd6dd38c063cdc53a1a8c

### Interface of Rocq / Coq where this bug occurs

_No response_

### Last version of Rocq / Coq where the bug did not occur

de8e6e97603ac2853a88438192821491a8b3163e

--- a/plugins/extraction/modutil.ml
+++ b/plugins/extraction/modutil.ml
@@ -117,15 +117,17 @@ let ind_iter_references do_term do_cons do_type ind =
   let cons_iter cp l = do_cons cp; List.iter type_iter l in
   let packet_iter i p =
     let () = do_type p.ip_typename_ref in
-    if lang () == Ocaml then
-      let inst = ind.ind_packets.(0).ip_typename_ref.inst in
-      (match ind.ind_equiv with
+    let () = if lang () == Ocaml then begin
+        let inst = ind.ind_packets.(0).ip_typename_ref.inst in
+        (match ind.ind_equiv with
          | Miniml.Equiv kne -> do_type { glob = (GlobRef.IndRef (MutInd.make1 kne, i)); inst };
-         | _ -> ());
+         | _ -> ())
+      end
+    in
     Array.iteri (fun j -> cons_iter p.ip_consnames_ref.(j)) p.ip_types
   in
-  if lang () == Ocaml then record_iter_references do_term ind.ind_kind;
-    Array.iteri (fun i p -> packet_iter i p) ind.ind_packets
+  let () = if lang () == Ocaml then record_iter_references do_term ind.ind_kind in
+  Array.iteri (fun i p -> packet_iter i p) ind.ind_packets
 
 let decl_iter_references do_term do_cons do_type =
   let type_iter = type_iter_references do_type

--- a/test-suite/output/bug_20711.out
+++ b/test-suite/output/bug_20711.out
@@ -1,0 +1,14 @@
+module Main where
+
+import qualified Prelude
+
+data Bool =
+   True
+ | False
+
+type Foo = Bool
+
+type Wrapper = Foo
+  -- singleton inductive, whose constructor was Build_Wrapper
+  
+

--- a/test-suite/output/bug_20711.v
+++ b/test-suite/output/bug_20711.v
@@ -1,0 +1,9 @@
+Require Extraction.
+
+Extraction Language Haskell.
+
+Definition foo := bool.
+
+Record Wrapper := { v1 : foo; }.
+
+Recursive Extraction Wrapper.

